File: wenger-feature/src/main/java/io/github/rcarlosdasilva/wenger/feature/region/RegionHandler.java
package io.github.rcarlosdasilva.wenger.feature.region;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import com.google.common.base.Strings;
import com.google.common.collect.*;
import io.github.rcarlosdasilva.kits.string.TextHelper;
import io.github.rcarlosdasilva.wenger.feature.config.AppProperties;
import org.springframework.beans.factory.SmartInitializingSingleton;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.util.*;
import java.util.stream.Collectors;

/**
 * 地区数据解析器
 *
 * @author Dean Zhao (rcarlosdasilva@qq.com)
 */
@ConditionalOnProperty(name = "app.misc.region.enable", havingValue = "true")
@Component
@EnableConfigurationProperties({AppProperties.class})
public class RegionHandler implements SmartInitializingSingleton {

  @Autowired
  private AppProperties appProperties;
  private final China china = new China();
  private BiMap<String, String> mapping;
  private String json;

  @Override
  public void afterSingletonsInstantiated() {
    String dataLocation = appProperties.getMisc().getRegion().getLocation();
    // 默认jar包内数据文件
    Resource resource = new ClassPathResource(dataLocation);

    // 尝试URL读取
    if (!resource.exists()) {
      try {
        resource = new UrlResource(dataLocation);
      } catch (MalformedURLException ex) {
        resource = null;
      }
    }

    // 尝试本地文件读取
    if (resource == null || !resource.exists()) {
      resource = new FileSystemResource(dataLocation);
    }

    // 无法读取
    if (!resource.exists()) {
      throw new RegionDataException("[地域] - 无法找到地域数据文件，请确保app.misc.china.location配置的文件地址存在");
    }

    Properties properties;
    try (InputStream is = resource.getInputStream();
         InputStreamReader isr = new InputStreamReader(is)) {
      properties = new Properties();
      properties.load(isr);
    } catch (IOException ex) {
      throw new RegionDataException("[地域] - 文件读取失败", ex);
    }

    // 解析
    BiMap<String, String> temMapping = HashBiMap.create();
    List<String> keys = new ArrayList<>(properties.stringPropertyNames());
    keys = ImmutableList.copyOf(Ordering.natural().sortedCopy(keys));
    Iterator<String> ki = keys.iterator();

    List<Province> temProvinces = Lists.newArrayList();
    Multimap<String, City> temCities = LinkedListMultimap.create();
    Multimap<String, District> temDistricts = LinkedListMultimap.create();

    // 归类
    while (ki.hasNext()) {
      String code = ki.next();
      String name = properties.getProperty(code);
      temMapping.forcePut(code, name);

      if (isProvince(code)) {
        temProvinces.add(new Province(code, name));
      } else if (isCity(code)) {
        temCities.put(fragmentProvince(code), new City(code, name));
      } else {
        temDistricts.put(fragmentCity(code), new District(code, name));
      }
    }

    // 连接省市区
    Map<String, Province> provinces = Maps.newLinkedHashMap();
    temProvinces.forEach(province -> {
      Collection<City> cc = temCities.get(province.getCode());
      Map<String, City> cities = Maps.newLinkedHashMap();

      cc.forEach(city -> {
        Collection<District> dc = temDistricts.get(city.getCode());
        Map<String, District> districts = dc.stream().collect(Collectors.toMap(District::getCode, district ->
            district, (ek, nk) -> ek, LinkedHashMap::new));

        city.setDistricts(ImmutableSortedMap.copyOf(districts));
        cities.put(city.getCode(), city);
      });

      province.setCities(ImmutableSortedMap.copyOf(cities));
      provinces.put(province.getCode(), province);
    });

    china.setProvinces(ImmutableSortedMap.copyOf(provinces));
    mapping = ImmutableBiMap.copyOf(temMapping);
  }

  private boolean isProvince(String code) {
    return code != null && code.matches(RegionConst.DEFAULT_PROVINCE_PATTERN);
  }

  private boolean isCity(String code) {
    return code != null && code.matches(RegionConst.DEFAULT_CITY_PATTERN);
  }

  private String fragmentProvince(String code) {
    return TextHelper.concat(TextHelper.sub(code, 0, 2), RegionConst.PROVINCE_SUFFIX);
  }

  private String fragmentCity(String code) {
    return TextHelper.concat(TextHelper.sub(code, 0, 4), RegionConst.CITY_SUFFIX);
  }

  /**
   * 将地区信息转成字符串并返回
   */
  public String toJSONString() {
    if (Strings.isNullOrEmpty(json)) {
      JSONObject root = new JSONObject();
      JSONArray jps = new JSONArray();
      root.put("provinces", jps);

      china.getProvinces().values().forEach(province -> {
        JSONObject jp = new JSONObject();
        JSONArray jcs = new JSONArray();
        jp.put("code", province.getCode());
        jp.put("name", province.getName());
        jp.put("cities", jcs);

        province.getCities().values().forEach(city -> {
          JSONObject jc = new JSONObject();
          JSONArray jds = new JSONArray();
          jc.put("code", city.getCode());
          jc.put("name", city.getName());
          jc.put("districts", jds);

          city.getDistricts().values().forEach(district -> {
            JSONObject jd = new JSONObject();
            jd.put("code", district.getCode());
            jd.put("name", district.getName());

            jds.add(jd);
          });

          jcs.add(jc);
        });

        jps.add(jp);
      });

      json = root.toJSONString();
    }
    return json;
  }

  /**
   * 根据code获取地区名
   *
   * @param code code
   * @return name
   */
  public Optional<String> getName(String code) {
    return code == null ? Optional.empty() : Optional.ofNullable(mapping.get(code));
  }

  /**
   * 根据地区名获取code，可能为null
   * <p>因为每个市下都有一个市辖区，所以不保证获取市辖区的code正确</p>
   *
   * @param name name
   * @return code
   */
  public Optional<String> getCode(String name) {
    return name == null ? Optional.empty() : Optional.ofNullable(mapping.inverse().get(name));
  }

  /**
   * 获取所有省
   *
   * @return 省
   */
  public Collection<Province> getProvinces() {
    return china.getProvinces().values();
  }

  /**
   * 根据code（不一定非要是省code，也可以是市或区code），获取所有市
   *
   * @param code code
   * @return 市
   */
  public Collection<City> getCities(String code) {
    Province province = china.getProvinces().get(fragmentProvince(code));
    if (province == null) {
      return Collections.emptyList();
    }

    return province.getCities().values();
  }

  /**
   * 根据code（市或区code），获取所有行政区
   *
   * @param code code
   * @return 行政区
   */
  public Collection<District> getDistricts(String code) {
    Province province = china.getProvinces().get(fragmentProvince(code));
    if (province == null) {
      return Collections.emptyList();
    }

    City city = province.getCities().get(fragmentCity(code));
    if (city == null) {
      return Collections.emptyList();
    }

    return city.getDistricts().values();
  }

  /**
   * 根据code，模糊取出code下所有省/市/区（code为null时返回所有省），如果code是行政区，返回null
   *
   * @param code code
   * @return data
   */
  public Collection<? extends AbstractRegion> getAmbiguousRegions(String code) {
    if (Strings.isNullOrEmpty(code)) {
      return getProvinces();
    } else if (isProvince(code)) {
      return getCities(code);
    } else if (isCity(code)) {
      return getDistricts(code);
    }

    return Collections.emptyList();
  }

  /**
   * 根据code，返回当前及上级所有地区名，例如：传入市的code，返回省与市
   *
   * @param code code
   * @return regions
   */
  public Collection<AbstractRegion> getFullRegion(String code) {
    List<AbstractRegion> result = Lists.newArrayList();

    Province province = china.getProvinces().get(fragmentProvince(code));
    if (province != null) {
      result.add(province);

      City city = province.getCities().get(fragmentCity(code));
      if (city != null) {
        result.add(city);

        District district = city.getDistricts().get(code);
        if (district != null) {
          result.add(district);
        }
      }
    }

    return result;
  }

  /**
   * 根据code，返回当前及上级所有的地区名称
   *
   * @param code code
   * @return names
   */
  public String getFullRegionName(String code) {
    return getFullRegion(code).stream().map(AbstractRegion::getName).collect(Collectors.joining());
  }

}

File: wenger-feature/src/main/java/io/github/rcarlosdasilva/wenger/feature/region/RegionHandler.kt
package io.github.rcarlosdasilva.wenger.feature.region

import com.alibaba.fastjson.JSONArray
import com.alibaba.fastjson.JSONObject
import com.google.common.collect.*
import io.github.rcarlosdasilva.kits.string.TextHelper
import io.github.rcarlosdasilva.wenger.common.exception.WengerRuntimeException
import io.github.rcarlosdasilva.wenger.feature.config.AppProperties
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.SmartInitializingSingleton
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.core.io.ClassPathResource
import org.springframework.core.io.FileSystemResource
import org.springframework.core.io.Resource
import org.springframework.core.io.UrlResource
import org.springframework.stereotype.Component
import java.io.IOException
import java.io.InputStreamReader
import java.net.MalformedURLException
import java.util.*

internal const val DEFAULT_PROVINCE_PATTERN = "^[1-9]\\d0000$"
internal const val DEFAULT_CITY_PATTERN = "^[1-9]\\d{3}00$"
internal const val DEFAULT_DISTRICT_PATTERN = "^\\d{4}(\\d[1-9]|[1-9]\\d)$"

/**
 * 地区数据解析器
 *
 * @author [Dean Zhao](mailto:rcarlosdasilva@qq.com)
 */
@ConditionalOnProperty(name = ["app.misc.region.enable"], havingValue = "true")
@Component
@EnableConfigurationProperties(value = [AppProperties::class])
class RegionHandler @Autowired constructor(
    private val appProperties: AppProperties
) : SmartInitializingSingleton {

  private val logger: Logger = LoggerFactory.getLogger(javaClass)

  private val china = Region("+86", "中国")
  private lateinit var indexes: BiMap<String, String> // 区域可逆键值索引，用于速查
  private lateinit var json: String

  override fun afterSingletonsInstantiated() {
    val resource = load()

    val properties = try {
      InputStreamReader(resource.inputStream).use {
        Properties().apply { this.load(it) }
      }
    } catch (ex: IOException) {
      throw WengerRegionException("[地域] - 文件读取失败", ex)
    }

    // 解析
    val tempProvinces = linkedMapOf<String, Region>()
    val tempCities = LinkedListMultimap.create<String, Region>()
    val tempDistricts = LinkedListMultimap.create<String, Region>()

    val tempIndexes = HashBiMap.create<String, String>()
    val keys = properties.stringPropertyNames().sorted()

    // 归类
    keys.forEach {
      val name = properties.getProperty(it)
      tempIndexes.forcePut(it, name)

      when {
        isProvince(it) -> tempProvinces[it] = Region(it, name)
        isCity(it) -> tempCities.put(fragment(it, 2), Region(it, name))
        isDistrict(it) -> tempDistricts.put(fragment(it, 4), Region(it, name))
        else -> logger.warn("[区域] - 有不合法数据：{}", it)
      }
    }

    // 连接省市区
    tempProvinces.forEach {
      val cities = tempCities.get(fragment(it.key, 2)) // it = 省code

      cities.forEach {
        val districts = tempDistricts.get(fragment(it.code, 4)) // it = 市Region
        it.subs = districts.associate { it.code to it }
      }
      it.value.subs = cities.associate { it.code to it }
    }

    china.subs = ImmutableSortedMap.copyOf(tempProvinces)
    indexes = ImmutableBiMap.copyOf(tempIndexes)

    toJson()
  }

  private fun load(): Resource {
    val dataLocation = appProperties.misc.region.location
    // 默认jar包内数据文件
    var resource: Resource = ClassPathResource(dataLocation)

    // 尝试URL读取
    if (!resource.exists()) {
      resource = try {
        UrlResource(dataLocation)
      } catch (ex: MalformedURLException) {
        FileSystemResource(dataLocation) // 尝试本地文件读取
      }
    }

    // 无法读取
    if (!resource.exists()) {
      throw WengerRegionException("[地域] - 无法找到地域数据文件，请确保app.misc.region.location配置的文件地址存在")
    }
    return resource
  }

  private fun toJson() {
    val root = JSONObject()
    val jps = JSONArray()
    root["provinces"] = jps

    china.subs!!.values.forEach { province ->
      val jp = JSONObject()
      val jcs = JSONArray()
      jp["code"] = province.code
      jp["name"] = province.name
      jp["cities"] = jcs

      province.subs!!.values.forEach { city ->
        val jc = JSONObject()
        val jds = JSONArray()
        jc["code"] = city.code
        jc["name"] = city.name
        jc["districts"] = jds

        city.subs!!.values.forEach { district ->
          val jd = JSONObject()
          jd["code"] = district.code
          jd["name"] = district.name
          jds.add(jd)
        }
        jcs.add(jc)
      }
      jps.add(jp)
    }

    json = root.toJSONString()
  }

  private fun fragment(code: String, length: Int): String = TextHelper.fill(code.substring(0, length), "0", -6)

  // ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓ functions ↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓

  fun isProvince(code: String): Boolean = code.matches(DEFAULT_PROVINCE_PATTERN.toRegex())
  fun isCity(code: String): Boolean = code.matches(DEFAULT_CITY_PATTERN.toRegex())
  fun isDistrict(code: String): Boolean = code.matches(DEFAULT_DISTRICT_PATTERN.toRegex())

  /**
   * 将地区信息转成字符串并返回
   */
  fun json(): String = json

  /**
   * 根据code获取地区名
   *
   * @param code code
   * @return name
   */
  fun name(code: String): Optional<String> = Optional.ofNullable(indexes[code])

  /**
   * 根据地区名获取code，可能为null
   *
   * 因为每个市下都有一个市辖区，所以不保证获取市辖区的code正确
   *
   * @param name name
   * @return code
   */
  fun code(name: String): Optional<String> = Optional.ofNullable(indexes.inverse()[name])

  /**
   * 获取所有省
   *
   * @return 省集合
   */
  fun getProvinces(): Collection<Region> = china.subs!!.values

  /**
   * 根据code（不一定非要是省code，也可以是市或区code），获取所有市。例如：370202为山东青岛市南区，返回山东下所有城市
   *
   * @param code code
   * @return 市
   */
  fun getCities(code: String): Collection<Region> =
      china.subs!![fragment(code, 2)]?.subs?.values ?: emptyList()

  /**
   * 根据code（市或区code），获取所有行政区
   *
   * @param code code
   * @return 行政区
   */
  fun getDistricts(code: String): Collection<Region> =
      china.subs!![fragment(code, 2)]?.subs!![fragment(code, 4)]?.subs?.values ?: emptyList()

  /**
   * 根据code，取出code下所有区域
   *
   * 即code=省时，取省下所有市，code=市时，取市下所有区，code=区时，返回空集合，code=null或空字符串时返回所有省，code不存在返回空集合
   *
   * @param code code
   * @return regions
   */
  fun getLower(code: String?): Collection<Region> {
    return when {
      code == null || code.isEmpty() -> getProvinces()
      isProvince(code) -> getCities(code)
      isCity(code) -> getDistricts(code)
      else -> emptyList()
    }
  }

  /**
   * 根据code，返回当前及上级所有地区名，例如：传入市的code，返回省与市
   *
   * 与getLower, getProvinces, getCities, getDistricts方法不同，返回的Region中无subs属性
   *
   * @param code code
   * @return regions
   */
  fun getFullRegion(code: String): Collection<Region> =
      listOf(fragment(code, 2), fragment(code, 4), code).mapNotNull {
        val name = indexes[it]
        if (name != null) Region(it, name) else null
      }

  /**
   * 根据code，返回当前及上级所有的地区名称
   *
   * 与getFullRegion类似，返回为全部地区的名称字符串
   *
   * @param code code
   * @param separator 区域之间间隔符
   * @return name
   */
  fun getFullRegionName(code: String, separator: String = ""): String =
      listOf(fragment(code, 2), fragment(code, 4), code).mapNotNull { indexes[it] }.joinToString(separator)

}

/**
 * 区域基本抽象类
 */
data class Region(val code: String, val name: String) {
  internal var subs: Map<String, Region>? = null
}

/**
 * 中国（区域根类）
 */
//class China

class WengerRegionException : WengerRuntimeException {
  constructor() : super()
  constructor(message: String?) : super(message)
  constructor(message: String?, cause: Throwable?) : super(message, cause)
  constructor(cause: Throwable?) : super(cause)
  constructor(message: String?, cause: Throwable?, enableSuppression: Boolean, writableStackTrace: Boolean) : super(message, cause, enableSuppression, writableStackTrace)
}

// TODO 数据Reload
// TODO 编辑数据，并回写文件